### PR TITLE
[`flake8-type-checking`] Avoid false positives for `|` in `TC008`

### PR DESCRIPTION
**Pull request by Daverball**
This is a follow-up to #15180

## Summary

`|` is not a valid runtime operation between types pre Python 3.10, so we should avoid unquoting the type expression.

Additionally most type checkers don't care about the execution context, so while they will allow `|` between types in stub files for convenience, they will still disallow it in type checking blocks, since those are treated as regular runtime code.

## Test Plan

`cargo nextest run`

### CODE REVIEW

**Comment by Daverball** [comment]
I've based this on the same branch as #15180 in order to reduce churn between the two changes, so there are some unrelated changes here until that has been merged.

**Comment by MichaReiser** [comment]
Reviewing this PR is a bit challenging right now because it's based on https://github.com/astral-sh/ruff/pull/15180 and I can't change the base because your an external contributor. I suggest we wait with reviewing until the base PR is merged. Please ping us if we happen to forget to review the PR after #15180 is merged.

**Comment by Daverball** [comment]
@MichaReiser @AlexWaygood You should be able to review this now

**Comment by Daverball** [comment]
The ecosystem results make sense to me now. Since airflow target Python 3.9+ mypy would flag those type aliases if they were unquoted. So a TC008 shouldn't trigger, despite there being no runtime effects.

**Comment by dangotbanned** [comment]
@Daverball 

I've got some false positives that seem to be this, in combination with being in a `TYPE_CHECKING` block:
- [workflow failure](https://github.com/vega/altair/actions/runs/12698005027/job/35395529522?pr=3747)
- Source
	- [1 & 3](https://github.com/vega/altair/blob/23b189103b8c5ac38928876cdde5ec5f71fcadc3/tools/schemapi/schemapi.py#L59)
	- [2](https://github.com/vega/altair/blob/23b189103b8c5ac38928876cdde5ec5f71fcadc3/tests/__init__.py#L24-L26)
- [`tool.ruff.target-version = "py39"`](https://github.com/vega/altair/blob/23b189103b8c5ac38928876cdde5ec5f71fcadc3/pyproject.toml#L206)

Should I open an issue for this?

**Comment by Daverball** [comment]
@dangotbanned Yes, that's this bug, it can also be hit in different contexts, it just previously wasn't as likely to be hit in `TYPE_CHECKING` blocks, because those also weren't handled properly yet.

Feel free to open an issue, if you think it makes it easier for others to discover that this issue is known and will be fixed.

**Comment by dangotbanned** [comment]
All good, thanks for clarifying @Daverball 🙂

**Comment by Daverball** [comment]
@MichaReiser @AlexWaygood Bumping this again, now that the 0.9 rush is over.

**Comment by MichaReiser** [comment]
Thanks for the ping. We decided that Alex should spend more time on Red Knot which leaves me as the main reviewer. I'll try to take a look but I'm not sure when I'll get to it because it also requires familiarizing myself with the type checking rules. Sorry for the long wait.

**Comment by Daverball** [comment]
@MichaReiser Alright, no worries. This PR should be fairly easy to review, even with surface knowledge, the only thing you need to know is that the union syntax `|` is only allowed to be used at runtime starting with Python 3.10, prior to that it will raise a `TypeError`, since `type` did not provide a `__or__` method that returns `UnionType`.

However type checkers still allow you to use the new syntax in older Python versions in non-runtime contexts like implicit/explicit forward references. But type checkers generally don't consider statements in `TYPE_CHECKING` blocks to be non-runtime. It is treated like any other code block, the only thing they do, is statically evaluate `TYPE_CHECKING` to `True` (and entirely skip analysis of code that isn't reachable this way).

There is a tiny bit of extra complexity for avoiding false negatives for `typing.Annotated`. This part should probably have been changed in #15180 in hindsight. All you need to know here is, that only the first argument to `typing.Annotated` is a type expression, so the rest of the arguments shouldn't influence our decision whether or not we can unquote.

TLDR: Annotated type alias value expressions always need to be considered in runtime context, unless they're explicitly wrapped in quotes. As a consequence, if a forward reference in the value expression uses `|`, we're not allowed to remove the quotes if the target python version is 3.9 or below.

**Review by MichaReiser — APPROVED** [review]
Thanks for the detailed explanation! It helped me to build the necessary context quickly to review the PR. I really appreciate it. 

Looks good and thanks again for taking the time to explain the changes in detail to me.
